### ISSUE TITLE
15b: Implement GP experiment with gplearn

### ISSUE THREAD

**Issue by ofekvs18** [ofekvs18/UnderGradProject#17]
**Parent**: #15
**Depends on**: #16 (shared eval code must be in utils first)

### Goal

Create `src/method3_gp.py` — the GP experiment script using gplearn's `SymbolicTransformer` with a custom multi-objective fitness function.

#### 1. Custom fitness function

Use `gplearn.fitness.make_fitness()` to create a combined AUC-ROC + AUC-PR metric:

```python
from gplearn.fitness import make_fitness
from sklearn.metrics import roc_auc_score, average_precision_score

def _auc_combined(y, y_pred, sample_weight):
    """
    Multi-objective fitness: weighted combination of AUC-ROC and AUC-PR.
    AUC-PR gets 2x weight since it's our primary metric.
    
    Returns a single scalar (higher = better).
    """
    # y_pred from gplearn is raw formula output, not probabilities
    # Handle NaN/Inf
    mask = np.isfinite(y_pred)
    if mask.sum() < 10 or y[mask].sum() < 3:
        return 0.0  # invalid formula
    y_clean = y[mask]
    pred_clean = y_pred[mask]
    
    try:
        auc_roc = roc_auc_score(y_clean, pred_clean)
        auc_pr = average_precision_score(y_clean, pred_clean)
    except Exception:
        return 0.0
    
    # Flip if inversely predictive
    if auc_roc < 0.5:
        auc_roc = 1.0 - auc_roc
        auc_pr = average_precision_score(y_clean, -pred_clean)
    
    # Weighted combination: 1/3 AUC-ROC + 2/3 AUC-PR
    return (auc_roc / 3) + (2 * auc_pr / 3)

auc_combined = make_fitness(function=_auc_combined, greater_is_better=True, wrap=False)
```

**Important**: `wrap=False` for faster single-threaded execution. If we need `n_jobs > 1`, change to `wrap=True`.

#### 2. GP configuration

Start small, with room to scale up:

```python
from gplearn.genetic import SymbolicTransformer

# Small run (Phase 1)
SMALL_CONFIG = dict(
    population_size=100,
    generations=20,
    tournament_size=10,
    function_set=['add', 'sub', 'mul', 'div', 'sqrt', 'log', 'abs'],
    metric=auc_combined,
    parsimony_coefficient=0.005,  # penalize complexity to prefer simpler formulas
    init_depth=(2, 4),           # start with simple formulas (2-4 features)
    p_crossover=0.7,
    p_subtree_mutation=0.1,
    p_hoist_mutation=0.05,       # fights bloat by shrinking programs
    p_point_mutation=0.1,
    p_point_replace=0.05,
    max_samples=0.8,             # subsample for diversity + OOB fitness
    feature_names=['hct', 'hgb', 'mch', 'mchc', 'mcv', 'plt', 'rbc', 'rdw', 'wbc'],
    hall_of_fame=20,             # keep top 20 programs
    n_components=10,             # output top 10 least-correlated programs
    verbose=1,
    random_state=42,
    n_jobs=1,
    warm_start=False,
)

# Medium run (Phase 2 — only if Phase 1 shows promise)
MEDIUM_CONFIG = {**SMALL_CONFIG, 'population_size': 300, 'generations': 50}

# Large run (Phase 3 — only if needed)
LARGE_CONFIG = {**SMALL_CONFIG, 'population_size': 500, 'generations': 100}
```

#### 3. Custom functions for gplearn

gplearn's built-in `'div'` already handles division by zero (returns 1.0). For `'sqrt'` and `'log'`, gplearn's built-in versions are also safe. We use the built-in set.

If the built-in `log` and `sqrt` don't exist, define custom safe versions:

```python
from gplearn.functions import make_function

def _safe_log(x):
    return np.log(np.abs(x) + 1e-6)

def _safe_sqrt(x):
    return np.sqrt(np.abs(x))

safe_log = make_function(function=_safe_log, name='log', arity=1, wrap=False)
safe_sqrt = make_function(function=_safe_sqrt, name='sqrt', arity=1, wrap=False)
```

Check gplearn docs at runtime to see which builtins are available.

#### 4. Script flow

```
1. Load data, get train/test splits
2. Fit SymbolicTransformer on train set with SMALL_CONFIG
3. Extract evolved programs (formula strings)
4. Evaluate ALL programs from hall_of_fame using evaluate_formula_full() from utils
5. Compare against baselines (Method 1, Method 2, LR)
6. Print top 10 by AUC-PR and by AUC-ROC
7. Save results and evolution curve
8. If promising → repeat with MEDIUM_CONFIG using warm_start=True
```

#### 5. Extracting formula strings from gplearn

```python
gp = SymbolicTransformer(**SMALL_CONFIG)
gp.fit(X_train, y_train)

# Extract programs
for i, program in enumerate(gp):
    formula_str = str(program)  # e.g., "add(div(rdw, hgb), mul(plt, rbc))"
    raw_fitness = program.raw_fitness_
    penalized_fitness = program.fitness_
    length = program.length_
    print(f"  Program {i}: fitness={raw_fitness:.4f}  length={length}  {formula_str}")
```

gplearn formulas use prefix notation like `add(rdw, hgb)`. We need to convert these to infix for our `evaluate_formula_full()`. Write a converter:

```python
def gplearn_to_infix(program_str):
    """Convert gplearn prefix notation to infix for eval().
    add(rdw, hgb) → (rdw + hgb)
    div(rdw, hgb) → (rdw / (hgb + 1e-6))
    """
    ...
```

OR: use gplearn's built-in `program.execute(X)` to get scores directly, bypassing string conversion. **Prefer this approach** — it avoids parsing issues and lets us feed the numpy scores straight to our AUC computation.

#### 6. Evolution tracking

Log per-generation stats for the experiment log:

```python
evolution_log = []
for gen in range(n_generations):
    # After each generation, record:
    evolution_log.append({
        'generation': gen,
        'best_fitness': ...,
        'avg_fitness': ...,
        'best_length': ...,
        'best_formula': ...,
    })
```

gplearn prints this with `verbose=1`, but we also want it in CSV.

### Output files

```
results/method3_gp/
├── all_programs.csv          # all hall-of-fame programs with metrics
├── top_formulas.csv          # top 10 by AUC-PR
├── evolution_curve.png       # fitness vs generation
├── top_pr_curves.png         # PR curves for top 5 GP formulas
├── auc_comparison.png        # bar chart: GP vs Method 1 vs Method 2 vs baseline
```

### Metrics to report (same as all methods)

For every evolved program: AUC-ROC, AUC-PR, precision, recall, F1, F2, P@R(0.25, 0.50, 0.75)

### Validation checks

```python
# V1: GP best should beat random chance
best_auc_roc = max(r['auc_roc'] for r in results)
assert best_auc_roc > 0.52, "GP didn't beat random"

# V2: No leakage
assert best_auc_roc < 0.95, f"AUC-ROC={best_auc_roc:.3f} — possible leakage"

# V3: Compare against Method 2 baseline
method2_best_pr = 0.0174
gp_best_pr = max(r['auc_pr'] for r in results)
print(f"GP best AUC-PR: {gp_best_pr:.4f} vs Method 2: {method2_best_pr:.4f}")
print(f"GP beats random formulas: {'YES' if gp_best_pr > method2_best_pr else 'NO'}")
```

### Dependencies

- `gplearn` (add to requirements.txt)
- `src/utils.py` (shared evaluation from #16)

### What NOT to do

- Do NOT use the test set during GP evolution (only for final evaluation)
- Do NOT run the large config unless the small config shows clear improvement over Method 2
- Do NOT use `n_jobs > 1` initially — gplearn parallel overhead often makes small runs slower
- Do NOT try to parse gplearn formula strings into eval-compatible format if `program.execute(X)` works — use the direct execution path
